File: src/main/java/io/github/tanguygab/tabadditions/shared/features/chat/ChatManager.java
package io.github.tanguygab.tabadditions.shared.features.chat;

import io.github.tanguygab.tabadditions.shared.TABAdditions;
import io.github.tanguygab.tabadditions.shared.TranslationFile;
import me.neznamy.tab.shared.TAB;
import me.neznamy.tab.shared.placeholders.PlayerPlaceholderImpl;
import me.neznamy.tab.shared.platform.TabPlayer;

import java.util.List;
import java.util.UUID;

public abstract class ChatManager {

    protected final TABAdditions plugin = TABAdditions.getInstance();
    protected final TAB tab = TAB.getInstance();
    protected final TranslationFile translation = plugin.getTranslation();
    protected final Chat chat;

    private final String data;
    protected final boolean toggleCmd;
    protected List<UUID> toggled;
    private PlayerPlaceholderImpl placeholder;
    private boolean invertPlaceholder = false;
    private String toggledOn;
    private String toggledOff;

    public ChatManager(Chat chat) {
        this(chat,false,null,null,null);
    }

    public ChatManager(Chat chat, boolean toggleCmd, String data, String cmd, String placeholder) {
        this.chat = chat;
        this.toggleCmd = toggleCmd;
        this.data = data;
        if (!toggleCmd) return;
        plugin.getPlatform().registerCommand(cmd);
        this.placeholder = tab.getPlaceholderManager().registerPlayerPlaceholder("%"+placeholder+"%",-1,p->hasCmdToggled((TabPlayer)p) ? "Off" : "On");
        toggled = plugin.loadData(data,true);
    }

    public void unload() {
        plugin.unloadData(data,toggled,toggleCmd);
    }

    protected void setToggleCmdMsgs(String toggledOn, String toggledOff) {
        this.toggledOn = toggledOn;
        this.toggledOff = toggledOff;
    }
    protected void invertToggleCmdPlaceholder() {
        invertPlaceholder = true;
    }
    protected boolean hasCmdToggled(TabPlayer p) {
        return p != null && toggled.contains(p.getUniqueId());
    }

    public boolean toggleCmd(TabPlayer player) {
        return plugin.toggleCmd(toggleCmd,player,toggled,placeholder,toggledOn,toggledOff,invertPlaceholder);
    }

    public abstract boolean onCommand(TabPlayer sender, String command);

}

File: src/main/kotlin/io/github/tanguygab/tabadditions/shared/features/chat/ChatManager.kt
package io.github.tanguygab.tabadditions.shared.features.chat

import io.github.tanguygab.tabadditions.shared.TABAdditions
import me.neznamy.tab.shared.TAB
import me.neznamy.tab.shared.placeholders.types.PlayerPlaceholderImpl
import me.neznamy.tab.shared.platform.TabPlayer
import java.util.UUID

abstract class ChatManager @JvmOverloads constructor(
    protected val chat: Chat,
    protected val toggleCmd: Boolean = false,
    private val data: String? = null,
    cmd: String? = null,
    placeholder: String? = null
) {
    protected val plugin = TABAdditions.INSTANCE
    protected val tab = TAB.getInstance()!!
    protected val translation = plugin.translation

    protected lateinit var toggled: MutableList<UUID>
    private lateinit var placeholder: PlayerPlaceholderImpl
    private var invertPlaceholder = false
    private var toggledOn: String? = null
    private var toggledOff: String? = null

    init {
        if (toggleCmd) {
            plugin.platform.registerCommand(cmd!!)
            this.placeholder = tab.placeholderManager.registerPlayerPlaceholder("%$placeholder%", -1)
            { if (hasCmdToggled(it as TabPlayer)) "Off" else "On" }
            chat.placeholders.add(this.placeholder)
            toggled = plugin.loadData(data!!, true)
        }
    }

    open fun unload() {
        plugin.unloadData(data!!, toggled, toggleCmd)
    }

    protected fun setToggleCmdMsgs(toggledOn: String, toggledOff: String) {
        this.toggledOn = toggledOn
        this.toggledOff = toggledOff
    }

    protected fun invertToggleCmdPlaceholder() {
        invertPlaceholder = true
    }

    fun hasCmdToggled(p: TabPlayer?) = p?.uniqueId in toggled

    fun toggleCmd(player: TabPlayer): Boolean {
        return plugin.toggleCmd(toggleCmd, player, toggled, placeholder, toggledOn!!, toggledOff!!, invertPlaceholder)
    }

    abstract fun onCommand(sender: TabPlayer, command: String): Boolean
}
